Add Percy visual regression testing (#2551)
alphagov/govuk-frontend: docs/releasing/testing-and-linting.md

# Testing and linting

> The CI lints SASS and JavaScript, runs unit and functional tests with Node, and generates snapshots for visual regression testing (Percy.io).

Reviewer: Will all our users know what 'visual regression testing' is? If not, do we need to link to an explanation?

Not sure about having 'Percy.io' in brackets at the end - there's no explanation of what it is (unless Percy.io = visual regression testing), and it isn't hyperlinked...maybe it would be ok to delete it from here, since we both hyperlink and explain it in line 53...

`npm test -- -u src/govuk/components/button`

This will update the snapshot file. Commit this file separately with a commit message that explains you're updating the snapshot file and an explanation of what caused the change.

> ## Visual regression testing with Percy

Reviewer 2: Thanks for adding these 🎉

We use [Percy](https://percy.io/), a visual regression testing tool, to generate and store screenshots of our components to help us check for any unintended visual changes.

We generate two screenshots for each default example of every component: one with JavaScript enabled and one with JavaScript disabled. Screenshots are not taken for all the different variations of each component. This tool is not a replacement for manual testing.

> The screenshots are public, so they can be checked without logging in. A BrowserStack account is needed to approve or reject any changes. It's the responsibility of the person reviewing the pull request code to approve any visual changes that Percy highlights.

Reviewer 3: I can access these when logged in, but am struggling to figure out how in an incognito window.

Created a dummy account for Github so I could view the Github Action logs and head off to the Percy build. Created a Browserstack account, but am getting "Access Denied" on the build at https://percy.io/55ba5e1a/govuk-frontend/builds/15872008

![Screenshot 2022-02-23 at 09 57 56](https://user-images.githubusercontent.com/22524634/155297056-137937d3-3b61-45ba-9b94-f6b825394022.png)

Author: @domoscargin Doh, hadn't enabled that setting. Do you want to give it a try now?

Reviewer 3: That works! 🎉  No login necessary.

`npm test -- -u src/govuk/components/button`

This will update the snapshot file. Commit this file separately with a commit message that explains you're updating the snapshot file and an explanation of what caused the change.

> ## Visual regression testing with Percy

Reviewer: @vanitabarrett Content looks really good! Left a few comments.

> We use [Percy](https://percy.io/), a visual regression testing tool, to generate and store screenshots of our components to help us check for any unintended visual changes.

Reviewer: ```suggestion
[Percy](https://percy.io/) is a visual regression testing tool. We use it to generate and store screenshots of our components to help us check for any unintended visual changes.
```

> We generate two screenshots for each default example of every component: one with JavaScript enabled and one with JavaScript disabled. Screenshots are not taken for all the different variations of each component. This tool is not a replacement for manual testing.

Reviewer: ```suggestion
We generate 2 screenshots for each default example of every component. One example has JavaScript enabled, the other has JavaScript disabled. Screenshots are not taken for all the different variations of each component. This tool is not a replacement for manual testing.
```

> The screenshots are public, so they can be checked without logging in. A BrowserStack account is needed to approve or reject any changes. It's the responsibility of the person reviewing the pull request code to approve any visual changes that Percy highlights.

Reviewer: ```suggestion
The screenshots are public, so you can check them without logging in. A BrowserStack account is needed to approve or reject any changes. If you're the reviewer of the pull request code, it's your responsibility to approve any visual changes Percy highlights.
```

Does the user need to have a BrowserStack account? If they do, do we need to link to BrowserStack signup?

Author: @EoinShaughnessy They need to have a BrowserStack account to approve or reject changes. I'm not sure about adding a link to BrowserStack signup - I think an existing member needs to add them to the Design System team. Maybe I'll add a line about asking the tech lead in the team to get you access...

> When running the tests locally (e.g: via `npm test`), Percy commands are ignored and no screenshots are generated. You will see the following message in your command line output: `[percy] Percy is not running, disabling snapshots`.

Reviewer: ```suggestion
When you run the tests locally (for example, using `npm test`), Percy commands are ignored and Percy does not generate any screenshots. You will see the following message in your command line output: `[percy] Percy is not running, disabling snapshots`.
```

### PRs from forks

> When Github Actions is running against a PR from a fork, the Percy secret is not available and therefore no screenshots are generated. Other tests will continue to run as normal. You will see the following messages in the output:

Reviewer: ```suggestion
When Github Actions is running against a PR from a fork, the Percy secret is not available and Percy does not generate any screenshots. Other tests will continue to run as normal. You will see the following messages in the output:
```

```
[percy] Skipping visual tests
[percy] Error: Missing Percy token
```

> This is the reason why we are unable to make Percy a required check for this repo. However, we should continue to act as if it is required. Percy runs should be approved before merging, and we should treat any failures as blocking.

Reviewer: ```suggestion
These messages are the reason we're unable to make Percy a required check for this repo. However, we should continue to act as if a Percy check is required. You should approve Percy runs before merging pull requests. Treat any failures as blocking.
```
Will it be obvious to users what action they should take in order to "treat any failures as blocking"?

Author: I guess what I mean there is, if the Percy build fails, don't just merge the PR because Github lets you. You should continue to review the changes and approve/reject them as necessary. There probably is a clearer way of writing that...

Reviewer: Thanks for explaining! Your wording looks good - maybe we could add in something like the following:

```suggestion
These messages are the reason we're unable to make Percy a required check for this repo. However, we should continue to act as if a Percy check is required. If the Percy build fails, do not merge the pull request even though GitHub would let you. Continue to review the changes, and approve or reject them as you would normally.
```

> Being unable to run Percy on PRs from forks is the reason why we are unable to make Percy a required check for this repo. However, we should continue to act as if a Percy check is required. If the Percy build fails, do not merge the pull request even though GitHub would let you. Continue to review the changes, and approve or reject them as you would normally.

Reviewer: ```suggestion
Being unable to run Percy on PRs from forks is the reason we're unable to make Percy a required check for this repo. However, we should continue to act as if a Percy check is required. If the Percy build fails, do not merge the pull request even though GitHub would let you. Continue to review the changes, and approve or reject them as you would normally.
```
